base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs

```diff
@@ -3132,59 +3133,78 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
+                    "[MIGRATIONS] v{}: No migration to perform for version network",
+                    migrate_from_version
+                );
                 continue;
             }
             // lets rewind to last known good accumulated difficulty so the db can be correctly calculated again
             rewind_to_height(db, last_correct_height)?;
         }
-        if migrate_from_version == 2 {
-            // Verify database consistency before starting migration
-            info!(target: LOG_TARGET, "Starting PayRef migration");
+
+        // MIGRATION: Add payref index, rebuild payref index to recover deleted payrefs, add reverse payref index
+        // Note: For previously running base nodes `last_migrated_version` was incremented beyond `MIGRATION_VERSION` up
+        //       to `migrate_from_version == 4`. This migration also fix the error introduced with the original payref
+        //       migration where it was only added for outputs in the unspent set, resulting in missing payrefs.
+        if (migrate_from_version == 2 || migrate_from_version == 3 || migrate_from_version == 4) && !payref_index_done {
+            info!(target: LOG_TARGET, "[MIGRATIONS] v{}: Starting PayRef migration", migrate_from_version);
 
             let read_txn = db.read_transaction()?;
-            let chain_height = match fetch_chain_height(&read_txn, &db.metadata_db) {
-                Ok(v) => v,
-                Err(_) => {
-                    // No chain data, skip PayRef rebuild
-                    drop(read_txn);
-                    continue;
-                },
-            };
+            let chain_height = fetch_chain_height(&read_txn, &db.metadata_db).unwrap_or(0);
             drop(read_txn);
-            // we need to clear the payref index first as they might now bw wrong
+
+            // Clear the payref index first as they might now be wrong
             {
                 let txn = db.write_transaction()?;
                 lmdb_clear(&txn, &db.payref_to_output_index)?;
                 txn.commit()?;
-                info!(target: LOG_TARGET, "Cleared PayRef index");
+                info!(target: LOG_TARGET, "[MIGRATIONS] v{}: Cleared PayRef index", migrate_from_version);
             }
+
             for height in 0..=chain_height {
+                if height % 100 == 0 {
+                    unsafe {
+                        LMDBStore::resize_if_required(&db.env, &db.env_config, None)?;
+                    }
+                }
                 process_payref_for_height(db, height)?;
             }
-            info!(target: LOG_TARGET, "PayRef index rebuild completed");
+
+            payref_index_done = true;
+            info!(target: LOG_TARGET, "[MIGRATIONS] v{}: PayRef index rebuild completed", migrate_from_version);
         }
-        // lets update the migration version
+
+        // Lets update the migration version
         {
+            let migrated_to_version = migrate_from_version + 1;
```

Comment:
```suggestion
            let migrated_to_version = migrate_from_version ;
```
was the +1 not a problem

Reply:
No, it was the combination of the `migrate_from_version` counter, where it ended and what is deemed as the version we have achieved - to make this clear I introduced the `migrated_to_version` variable and changed `last_migrated_version..=MIGRATION_VERSION` to `last_migrated_version..MIGRATION_VERSION`. 

If we use `last_migrated_version..=MIGRATION_VERSION` the start and end logic is not 100%  as it would revisit the loop if we are on v4 and we should be on v4 - now it will only go into the loop if a migration is needed.